linera-client/src/chain_listener.rs

```diff
@@ -274,37 +199,143 @@ impl<C: ClientContext> ChainClientListener<C> {
         if new_chains.is_empty() {
             return Ok(());
         }
-        let mut context_guard = self.listener.context.lock().await;
-        let timestamp = block.header.timestamp;
+        let mut new_ids = BTreeSet::new();
+        let mut context_guard = self.context.lock().await;
         for (new_id, owners) in new_chains {
             let key_pair = owners
                 .iter()
                 .find_map(|owner| context_guard.wallet().key_pair_for_owner(owner));
             if key_pair.is_some() {
                 context_guard
-                    .update_wallet_for_new_chain(*new_id, key_pair, timestamp)
+                    .update_wallet_for_new_chain(*new_id, key_pair, block.header.timestamp)
                     .await?;
-                self.listener.run_with_chain_id(*new_id);
+                new_ids.insert(*new_id);
+            }
+        }
+        drop(context_guard);
+        for new_id in new_ids {
+            self.listen(new_id).await?;
+        }
+        Ok(())
+    }
+
+    /// Processes the inboxes of all chains that are subscribed to streams with new events.
+    async fn process_new_events(&mut self, chain_id: ChainId) -> Result<(), Error> {
+        // TODO(#365): Support subscriptions to chains other than the admin chain.
+        let admin_id = self.context.lock().await.wallet().genesis_admin_chain();
+        if chain_id == admin_id {
+            let chain_ids = self.listening.keys().cloned().collect::<Vec<_>>();
+            for chain_id in chain_ids {
+                if chain_id != admin_id {
+                    self.maybe_process_inbox(chain_id).await?;
+                }
             }
         }
         Ok(())
     }
 
+    /// Start listening for notifications about the given chain.
+    async fn listen(&mut self, chain_id: ChainId) -> Result<(), Error> {
```

Comment:
Can we dynamically add new chains to listen to? i.e. can we call `listen` from the outisde and this process will add new subscription streams after the `self.run` has been called?

Reply:
To do that, we'll have to add a third arm to the `select!` statement that notifies about a new chain ID to listen to. Should be doable if we need it.